tmexp/create_bow.py

```diff
@@ -0,0 +1,209 @@
+    docs_output_path = os.path.join(output_dir, "docs." + dataset_name + ".txt")
+    check_remove_file(docs_output_path, logger, force)
+    check_fraction(min_word_frac, "min-word-frac")
+    check_fraction(max_word_frac, "max-word-frac")
+
+    logger.info("Reading pickled dict from '%s' ..." % input_path)
+    with open(input_path, "rb") as fin:
+        input_dict = pickle.load(fin)
+
+    logger.info("Computing bag of words ...")
+    langs = create_language_list(langs, exclude_langs)
+    all_refs = sorted(input_dict["files_info"])
+    doc_freq: Counter = Counter()
+    bow: Dict[str, Dict[int, Dict[str, Any]]] = {}
+    num_docwords = 0
+    num_blobs = 0
+    for file_path, blob_dict in tqdm.tqdm(input_dict["files_content"].items()):
+        refs = sorted(
+            ref
+            for ref, file_dict in input_dict["files_info"].items()
+            if file_path in file_dict
+        )
+        lang = input_dict["files_info"][refs[0]][file_path]["language"]
+        if lang not in langs:
+            continue
+        word_counts = {}
+        for blob_hash, feature_dict in blob_dict.items():
+            word_dict: Counter = Counter()
+            for feature in features:
+                if feature in feature_dict:
+                    word_dict.update(feature_dict[feature])
+            if word_dict:
+                num_blobs += 1
+                doc_freq.update(word_dict.keys())
+                word_counts[blob_hash] = word_dict
+        if not word_counts:
+            continue
+        file_bows: Dict[int, Dict[str, Any]] = {}
+        prev_ref = None
+        prev_blob = None
+        if topic_model == DIFF_MODEL:
+            prev_word_counts: Counter = Counter()
+        for ref in all_refs:
+            if ref in refs:
+                cur_blob = input_dict["files_info"][ref][file_path]["blob_hash"]
+                cur_word_counts = copy.deepcopy(word_counts[cur_blob])
+            else:
+                if prev_ref is None:
+                    continue
+                cur_blob = None
+                cur_word_counts = Counter()
+            if not (cur_blob is None and topic_model == HALL_MODEL):
+                prev_id = len(file_bows) - 1
+                if prev_blob == cur_blob:
+                    file_bows[prev_id]["refs"].append(ref)
+                else:
+                    file_bows[prev_id + 1] = {"refs": [ref]}
+                    if topic_model == HALL_MODEL:
+                        file_bows[prev_id + 1]["all"] = cur_word_counts
+                    elif topic_model == DIFF_MODEL:
+                        cur_word_counts.subtract(prev_word_counts)
+                        file_bows[prev_id + 1]["added"] = {
+                            word: count
+                            for word, count in cur_word_counts.items()
+                            if count > 0
+                        }
+                        file_bows[prev_id + 1]["removed"] = {
```

Comment:
I know they use this in the diff model (deleted words), but I think it introduces duplication. We should probably have a flag to consider only additions, wdyt?

Reply:
I am not sure it's bad, because the deleted parts help us to remove topics as well. If you look at how he computes the final topics distrib for each version it makes sense. However we could experiment, as you want

Reply:
True, let's keep it